### PR TITLE
[DROOLS-4070] allow to define the priority of alternative services du…

### PR DESCRIPTION
**Pull request by mariofusco**

### CODE REVIEW

**Comment by mariofusco** [comment]
jenkins execute fdb

**Comment by mariofusco** [comment]
jenkins execute fdb

File: kie-api/src/main/java/org/kie/api/internal/utils/ServiceDiscoveryImpl.java [hunk 1]
```diff
@@ -126,7 +113,12 @@ private void processKieService(ClassLoader classLoader, String key, String value
                     childServices.computeIfAbsent( serviceName, k -> new ArrayList<>() )
                             .add( newInstance( classLoader, value.substring( 1 ) ) );
                 } else {
-                    services.put( serviceName, newInstance( classLoader, value ) );
+                    int priority = 0;
+                    if ( value.charAt( 0 ) >= '0' && value.charAt( 0 ) <= '9' ) {
```
Comment (lucamolteni) [inline kie-api/src/main/java/org/kie/api/internal/utils/ServiceDiscoveryImpl.java:117]: extract value.charAt(0)

Also what does this function do? it's unreadable IMHO

File: kie-api/src/test/resources/META-INF/kie.conf.test1 [hunk 2]
```diff
@@ -0,0 +1 @@
+org.kie.api.internal.assembler.KieAssemblers = 3org.kie.api.internal.utils.MockAssemblersImpl
```
Comment (lucamolteni) [inline kie-api/src/test/resources/META-INF/kie.conf.test1:1]: I'm not sure I like the number before the component.

Why not using the marker annotation on the implementation as we discussed?

**Review by lucamolteni — CHANGES_REQUESTED** [review]

**Comment by gitgabrio** [comment]
@mariofusco @lucamolteni @danielezonca 
Hi. I understand the needs for this "prioritization" mechanism. But I think that "spreading" it all over the codebase (prepending a number to kie.conf, as in this proposal, or adding annotation on every concrete implementation) is pretty cumbersonme to maintain in the long term. It is based on the assumption - that I think is wrong - that the codebase is fixed, and that anyone working on it (now and in the future) knows exactly every place where this is used.
Another side effect would be that changing the priority, or inserting a new implementation, would require to modify all the definitions of implementation with lower priority. 
I would prefer to gather this information in one single file (class? properties ?) where each known implementation is listed with its priority.
That way
1) there is a clear and easy way to know the priority of every given implementation
2) changing/removing/adding implementation (with consequent priority) would require modification of only that file
3) in the unlucky situation of two concurrent modification (by different developers) on the priority of the same service, a merge conflict will appear immediately, and not only at runtime

**Comment by mariofusco** [comment]
jenkins execute fdb

**Review by lucamolteni — APPROVED** [review]

File: kie-api/src/test/java/org/kie/api/internal/utils/AnotherMockAssemblersImpl.java [hunk 3]
```diff
@@ -14,5 +14,6 @@
 
 package org.kie.api.internal.utils;
 
+@ServicePriority( 4 )
```
Comment (gitgabrio) [inline kie-api/src/test/java/org/kie/api/internal/utils/AnotherMockAssemblersImpl.java:17]: @mariofusco 
I strongly think we should not do this kind of stuff, as I tried to explain.

File: kie-api/src/test/java/org/kie/api/internal/utils/MockAssemblersImpl.java [hunk 4]
```diff
@@ -28,6 +28,7 @@
 import org.kie.api.io.ResourceType;
 import org.kie.api.io.ResourceWithConfiguration;
 
+@ServicePriority( 3 )
```
Comment (gitgabrio) [inline kie-api/src/test/java/org/kie/api/internal/utils/MockAssemblersImpl.java:31]: @mariofusco 
see above

File: kie-api/src/main/java/org/kie/api/internal/utils/ServiceDiscoveryImpl.java [hunk 5]
```diff
@@ -140,14 +141,16 @@ private void processKieService(ClassLoader classLoader, String key, String value
         }
     }
 
-    private Map<String, Object> buildMap() {
-        Map<String, Object> servicesMap = new HashMap<>();
-        for (Map.Entry<String, Object> serviceEntry : services.entrySet()) {
+    private Map<String, List<Object>> buildMap() {
```
Comment (gitgabrio) [inline kie-api/src/main/java/org/kie/api/internal/utils/ServiceDiscoveryImpl.java:144]: @mariofusco 
All that PR resolves to the population of that map.
I think we should put the required information in one single place, easy to find and modify.
Polluting the code with "priority information" scattered all over the codebase just add unneeded complexity and decrease the overall maintainability, without a real benefit.

**Review by gitgabrio — CHANGES_REQUESTED** [review]
@mariofusco 
Please, change the way of "priority mapping"

**Comment by mariofusco** [comment]
Hi @gitgabrio and thanks for your feedback. I agree that having a scattered definition of service implementations' priorities may eventually cause some maintenance problems and challenges. 

That said, I discussed this with @lucamolteni and @danielezonca and it seems we agreed (correct me if I'm wrong) that having a centralized configuration file will be equally impractical (with your suggestion every time a new service is added to a new or an existing module it will be also necessary to edit this single configuration file in kie-api) and will also defeat the purpose of an automatic and dynamic service discovery mechanism.

**Comment by mariofusco** [comment]
jenkins execute fdb

**Comment by gitgabrio** [comment]
@mariofusco @lucamolteni @danielezonca 
Sorry, but I strongly disagree.
Behind the scenes, in a nutshell, what you are actually doing is to bind a "name" (the implementation class name) with a "number" (the priority).
For this mapping to be meaningful, thought, you have to gather all those information in one place (the "Map") you create.
So, you are
1) scattering all this priority level all over the codebase
2) collect them at runtime
3) create a map out of them, also at runtime.
That process - actually - seems a really original way to transform a good ol' "property" file (key/value), boringly easy to read/write and absolutely devoid of fascinating unexpected events, in something extremely thrilling to discover at runtime: work or not? What will happen ? :smile: 

_"equally impractical (with your suggestion every time a new service is added to a new or an existing module it will be also necessary to edit this single configuration file in kie-api)"_: but with this implementation you will have to change _all/most_ of the files (i.e. all the annotation of the classes whose priority is impacted), and not only one!

_" defeat the purpose of an automatic and dynamic service discovery mechanism."_ those priorities are not dynamic at all - they are hard-coded. And they work only if there is a careful attention to not overwrite the same priority level: all of this - IMHO - is pretty  contradictory for a "automatic and dynamic service discovery mechanism"

**Comment by gitgabrio** [comment]
@mariofusco @danielezonca @lucamolteni 
Following private conversationx, I realized that I did not explain the root reasons of my remarks.
I think prioritization of services is a feature that could be useful for different uses cases.
One of them is the following one:
1) in the near future there will be two or three PMML implementations, 
2) I'm sure (somehow it has already been anticipated) that the final user will need an easy way to switch between them without recompiling drools and possibly without messing too much with the pom; 
3) moreover, different users may choose different implementations, regardless of our "default". 
Currently I'm managing that with a system property, but the prioritization mechanism could suit the needs if easily configurable by end user.

Now, I forgot the fact that "kie.conf" could be written by final user in his own project, overriding the ones written inside drools.
Being that the case, I agree on resolve the prioritization based on the kie.conf (it would be impossible with the annotation).

Thx! :+1:

**Comment by mariofusco** [comment]
@gitgabrio @lucamolteni @danielezonca I implemented all the changes that we have discussed, please give it one last review.

File: kie-api/src/main/java/org/kie/api/internal/utils/ServiceDiscoveryImpl.java [hunk 6]
```diff
@@ -112,8 +100,10 @@ public void registerConfs( ClassLoader classLoader, URL url ) {
                     processKieService( classLoader, entry[0].trim(), entry[1].trim() );
                 }
             }
-        } catch (Exception exc) {
-            throw new RuntimeException( "Unable to build kie service url = " + url.toExternalForm(), exc );
+        } catch (RuntimeException re) {
+            throw re;
```
Comment (danielezonca) [inline kie-api/src/main/java/org/kie/api/internal/utils/ServiceDiscoveryImpl.java:104]: What is the reason of this change? I understand that this avoid to wrap the exception but I think it makes more sense to add some contextual information like in the other catch (`"Unable to build kie service url = " + url.toExternalForm()`).
Wdyt?

File: kie-api/src/test/java/org/kie/api/internal/utils/ServiceDiscoveryImplTest.java [hunk 7]
```diff
@@ -16,30 +16,77 @@
...
+        ClassLoader cl = ServiceDiscoveryImplTest.class.getClassLoader();
+
+        try {
+            serviceDiscovery.registerConfs( cl, getUrl( cl, "META-INF/kie.conf.test1" ) );
+            serviceDiscovery.registerConfs( cl, getUrl( cl, "META-INF/kie.conf.test2" ) );
+            serviceDiscovery.getServices();
+            fail( "Trying to load a duplicated service should fail" );
+        } catch (Exception e) {
+            System.out.println( e.getMessage() );
+        }
```
Comment (danielezonca) [inline kie-api/src/test/java/org/kie/api/internal/utils/ServiceDiscoveryImplTest.java:66]: Why this `try catch` instead of just `throws Exception`?

File: kie-api/src/main/java/org/kie/api/internal/utils/ServiceDiscoveryImpl.java [hunk 8]
```diff
@@ -126,33 +116,56 @@ private void processKieService(ClassLoader classLoader, String key, String value
...
             } catch (RuntimeException e) {
                 if (optional) {
                     log.info("Cannot load service: " + serviceName);
                 } else {
-                    System.out.println("Loading failed because " + e.getMessage());
+                    log.error("Loading failed because " + e.getMessage());
                     throw e;
                 }
             }
-            log.debug( "Adding Service {}\n", value );
```
Comment (danielezonca) [inline kie-api/src/main/java/org/kie/api/internal/utils/ServiceDiscoveryImpl.java:139]: Can you please restore this log or add a log message inside the [`if` block](https://github.com/kiegroup/droolsjbpm-knowledge/blob/03ba485e72b2ba1796765f4e2f9e68e11bf1b2de/kie-api/src/main/java/org/kie/api/internal/utils/ServiceDiscoveryImpl.java#L115-L117) too?

File: kie-api/src/main/java/org/kie/api/internal/utils/ServiceDiscoveryImpl.java [hunk 9]
```diff
@@ -126,33 +116,56 @@ private void processKieService(ClassLoader classLoader, String key, String value
...
+                    }
+                }
+            }
+        }
+
+        if (!childServices.isEmpty()) {
+            throw new RuntimeException("Child services " + childServices.keySet() + " have no parent");
+        }
+
+        return servicesMap;
```
Comment (danielezonca) [inline kie-api/src/main/java/org/kie/api/internal/utils/ServiceDiscoveryImpl.java:166]: Can you please log as `debug` or `trace` the whole content of the map (with one or more lines of log)? So that it will be easy to get the full status of priorities and services.
It is fine for me also inside `getServices` method

File: kie-api/src/main/java/org/kie/api/internal/utils/ServiceDiscoveryImpl.java [hunk 10]
```diff
@@ -116,9 +116,12 @@ private void processKieService(ClassLoader classLoader, String key, String value
                     childServices.computeIfAbsent( serviceName, k -> new ArrayList<>() )
                             .add( newInstance( classLoader, value.substring( 1 ) ) );
                 } else {
-                    Object service = newInstance( classLoader, value );
-                    ServicePriority servicePriority = service.getClass().getAnnotation( ServicePriority.class );
-                    int priority = servicePriority != null ? servicePriority.value() : 0;
+                    int priority = 0;
+                    int separatorPos = value.indexOf( ';' );
+                    if (separatorPos > 0) {
+                        priority = Integer.parseInt( value.substring( separatorPos+1 ).trim() );
```
Comment (gitgabrio) [inline kie-api/src/main/java/org/kie/api/internal/utils/ServiceDiscoveryImpl.java:122]: @mariofusco 
This may throw a couple of exception: NPE and NumberFormat exception. Maybe using 
public String[] split(String regex) 
remove one of them...

File: kie-api/src/test/java/org/kie/api/internal/utils/ServiceDiscoveryImplTest.java (hunk 7 above)
Comment (mariofusco) [reply kie-api/src/test/java/org/kie/api/internal/utils/ServiceDiscoveryImplTest.java:66]: If I throw the Exception the test will fail, while in that case the throwing of an Exception is the expected behaviour and indeed the test fail if the Exception is NOT thrown.

Comment (gitgabrio) [reply kie-api/src/test/java/org/kie/api/internal/utils/ServiceDiscoveryImplTest.java:66]: @mariofusco 
being that the case, would it be possible to use the form

`@Test(expected = Exception.class)` (better to use a more specific exception)

?
This is what I'm using in PMML to make clear that I'm expecting a test to throw an exception - eventually even a specific Exception (my code almost always throw specific exception, so a generic/unexpected one should be marked as a test failure)

File: kie-api/src/main/java/org/kie/api/internal/utils/ServiceDiscoveryImpl.java (hunk 10 above)
Comment (mariofusco) [reply kie-api/src/main/java/org/kie/api/internal/utils/ServiceDiscoveryImpl.java:121]: @gitgabrio I added one more check to be sure that the ';' is not the last char of the string and in that case I just ignore it. Regarding the NFE it's ok for me to let it be thrown because it means that user put in an invalid value for priority and there isn't any form of recovery that I could do from that error.

File: kie-api/src/main/java/org/kie/api/internal/utils/ServiceDiscoveryImpl.java [hunk 11]
```diff
@@ -115,10 +113,11 @@ private void processKieService(ClassLoader classLoader, String key, String value
                 if ( value.startsWith( "+" ) ) {
                     childServices.computeIfAbsent( serviceName, k -> new ArrayList<>() )
                             .add( newInstance( classLoader, value.substring( 1 ) ) );
+                    log.debug( "Added child Service " + value );
                 } else {
                     int priority = 0;
                     int separatorPos = value.indexOf( ';' );
-                    if (separatorPos > 0) {
+                    if (separatorPos > 0 && separatorPos < value.length()-1) {
```
Comment (gitgabrio) [inline kie-api/src/main/java/org/kie/api/internal/utils/ServiceDiscoveryImpl.java:120]: @mariofusco 
Don't you think this 
`
 int separatorPos = value.indexOf( ';' );
 if (separatorPos > 0 && separatorPos < value.length()-1) {`

is a somehow harder to read (two lines of code plus an  _and_ operator) then

`String[] parts = value.split(";")` ?

Out of curiosity, is there any technical/performance reason to prefer the former to the latter ?

Comment (mariofusco) [reply kie-api/src/main/java/org/kie/api/internal/utils/ServiceDiscoveryImpl.java:120]: @gitgabrio I find the 2 options equally readable, but readability is in the eyes of the reader. It's true that there's a small performance cost in using the split, but this is not my main concern here. In general when I know that there are (at most) only 2 parts I use the indexOf while I use the split when there can be 3 or more. In this specific case this also allow me to throw a NFE also in the case when the user write something like mypkg.MyClass;3;4 because it will try to parse "3;4" as priority. If I use the split it will take '3' as priority (unless I don't add a further check to make sure that there at most 2 parts) that is not what I want here.

Comment (lucamolteni) [reply kie-api/src/main/java/org/kie/api/internal/utils/ServiceDiscoveryImpl.java:120]: `mypkg.MyClass;3;4 ` is an invalid syntax so you should throw an exception if the array is more than two elements

And yes, `value.split` is far more readable than indexOf

Comment (gitgabrio) [reply kie-api/src/main/java/org/kie/api/internal/utils/ServiceDiscoveryImpl.java:120]: @mariofusco @lucamolteni 
I know discussing so much of three lines of code seems too much.
The reason I insist on that, though, it is because IMO behind the scene this modification imply a radical change in the paradigm/semantic of the kie.conf itself, for which the "split" (o something analogue) would be much more suited.
I'll be glad to explain better this point with a personal conversation.

Beside that, as Luca somehow pointed out, we should identify different Exception condition and manage them appropriately: "mypkg.MyClass;3;4" is a different error then "mypkg.MyClass;3,4" or "mypkg.MyClass;3.4" or ""mypkg.MyClass;a": and I can't see a "nice" way to do that with the "valueOf" approach.

File: kie-api/src/main/java/org/kie/api/internal/utils/ServiceDiscoveryImpl.java (hunk 9 above)
Comment (tarilabs) [reply kie-api/src/main/java/org/kie/api/internal/utils/ServiceDiscoveryImpl.java:174]: +1

**Review by tarilabs — APPROVED** [review]
looks good to me, with one minor suggestion I concur with Daniele

File: kie-api/src/main/java/org/kie/api/internal/utils/ServiceDiscoveryImpl.java (hunk 9 above)
Comment (tarilabs) [reply kie-api/src/main/java/org/kie/api/internal/utils/ServiceDiscoveryImpl.java:174]: nevermind my previous comment, I understand is performed by https://github.com/kiegroup/droolsjbpm-knowledge/pull/466/files#diff-adbb6a46fc107d7f6590434788b73ac5R164-R172 as projected on each service. Thanks!
